### PR TITLE
Replace deprecated actor publisher/subscriber

### PR DESCRIPTION
**Pull request by notxcain**
This PR (ref #32) removes the usage of deprecated `AkkaSink.actorSubscriber` and `AkkaSource.actorPublisher` in favour of `AkkaSink.queue` and `AkkaSource.queue` correspondently.

@krasserm I was wrong about this change being non-breaking as it adds `Concurrent` constraint to some methods to be able to use `Stream#interruptWhen`.

@daenyth I would be very glad if you could review the implementation of `publisherStream`, it looks a bit clumsy to me, all tests pass though.

### CODE REVIEW

File: streamz-converter/src/main/scala/streamz/converter/Converter.scala [hunk 1]
```diff
@@ -133,34 +130,31 @@ trait Converter {
     }).mapMaterializedValue(_ => NotUsed)
   }
 
-  private def subscriberStream[F[_], A](subscriber: ActorRef)(implicit context: ContextShift[F], F: Async[F]): Stream[F, A] = {
-    val pull = context.shift >> F.async((callback: Callback[Option[A]]) => subscriber ! Request(callback))
+  private def subscriberStream[F[_], A](subscriber: SinkQueueWithCancel[A])(implicit context: ContextShift[F], F: Async[F]): Stream[F, A] = {
+    val pull = context.shift >> F.liftIO(IO.fromFuture(IO(subscriber.pull())))
     Stream.repeatEval(pull).unNoneTerminate
-      .onFinalize {
```
Comment (Daenyth) [inline streamz-converter/src/main/scala/streamz/converter/Converter.scala]: It looks like `onFinalize` is still there - is there a reason not to cancel the subscriber here?

Comment (notxcain) [reply streamz-converter/src/main/scala/streamz/converter/Converter.scala]: I think you’re right. It should be canceled here.

**Comment by notxcain** [comment]
@daenyth there’re no other breaking changes.

File: streamz-converter/src/main/scala/streamz/converter/Converter.scala [hunk 2]
```diff
@@ -133,34 +130,31 @@ trait Converter {
...
 
-  private def publisherStream[F[_], A](publisher: ActorRef, stream: Stream[F, A])(implicit context: ContextShift[F], F: Async[F]): Stream[F, Unit] = {
-    def publish(i: A): F[Option[Unit]] = context.shift >> F.async[Option[Unit]](callback => publisher ! Next(i, callback))
-    stream.evalMap(publish).unNoneTerminate
+  private def publisherStream[F[_], A](publisher: SourceQueueWithComplete[A], stream: Stream[F, A])(implicit context: ContextShift[F], F: Concurrent[F]): Stream[F, Unit] = {
+    def publish(a: A): F[Option[Unit]] = F.liftIO(IO.fromFuture(IO(publisher.offer(a)))).flatMap {
+      case QueueOfferResult.Enqueued => ().some.pure[F]
+      case QueueOfferResult.Failure(cause) => F.raiseError[Option[Unit]](cause)
+      case QueueOfferResult.QueueClosed => none[Unit].pure[F]
+      case QueueOfferResult.Dropped => F.raiseError[Option[Unit]](new IllegalArgumentException("Should never happen"))
```
Comment (krasserm) [inline streamz-converter/src/main/scala/streamz/converter/Converter.scala]: This can happen when if downstream is terminated.

Comment (krasserm) [reply streamz-converter/src/main/scala/streamz/converter/Converter.scala]: Nevermind my previous comment.

**Comment by krasserm** [comment]
LGTM, thanks for the PR! Will merge after subscriber cancellation is done in `onFinalize`.

File: streamz-converter/src/main/scala/streamz/converter/Converter.scala [hunk 3]
```diff
@@ -133,34 +130,31 @@ trait Converter {
     }).mapMaterializedValue(_ => NotUsed)
   }
 
-  private def subscriberStream[F[_], A](subscriber: ActorRef)(implicit context: ContextShift[F], F: Async[F]): Stream[F, A] = {
-    val pull = context.shift >> F.async((callback: Callback[Option[A]]) => subscriber ! Request(callback))
+  private def subscriberStream[F[_], A](subscriber: SinkQueueWithCancel[A])(implicit context: ContextShift[F], F: Async[F]): Stream[F, A] = {
+    val pull = context.shift >> F.liftIO(IO.fromFuture(IO(subscriber.pull())))
```
Comment (notxcain) [inline streamz-converter/src/main/scala/streamz/converter/Converter.scala]: @Daenyth @krasserm Do we really need `context.shift` here?

File: streamz-converter/src/main/scala/streamz/converter/Converter.scala (hunk 2 above)
Comment (notxcain) [reply streamz-converter/src/main/scala/streamz/converter/Converter.scala]: I've clarified this part.

**Comment by Daenyth** [comment]
Shift seems right to me

File: streamz-converter/src/main/scala/streamz/converter/Converter.scala (hunk 1 above)
Comment (notxcain) [reply streamz-converter/src/main/scala/streamz/converter/Converter.scala]: Fixed, but now I wonder is it really necessary? `SinkQueueWithCancel#pull` fails only if stream is failed. `Stream.repeatEval(pull).unNoneTerminate` completes only when `SinkQueueWithCancel#pull` returns `None`, i.e. stream completes.

**Comment by notxcain** [comment]
@Daenyth should it be added to [publish](https://github.com/krasserm/streamz/pull/52/files#diff-830512056ff77de8e13e288ce944f772R135) as well?

File: streamz-converter/src/main/scala/streamz/converter/Converter.scala (hunk 1 above)
Comment (Daenyth) [reply streamz-converter/src/main/scala/streamz/converter/Converter.scala]: It could be cancelled too though

**Comment by Daenyth** [comment]
That seems reasonable. @SystemFw might be able to say better than me

**Comment by notxcain** [comment]
Have you seen this [PR](https://github.com/functional-streams-for-scala/fs2/pull/1259)?

**Comment by Daenyth** [comment]
It was a separate repo before. I've used it, works fine AFAIK

**Comment by krasserm** [comment]
@notxcain @Daenyth what do you think of a complete re-implementation of the converters on top of https://github.com/functional-streams-for-scala/fs2/pull/1259 ?

**Comment by notxcain** [comment]
@krasserm I think it's very reasonable.

**Comment by Daenyth** [comment]
It's possible. Do you have benchmarks? It feels like it would be going through multiple layers of indirection when it doesn't necessarily need that. I also wonder about managing the `Mat` values - it seems like you'd have to convert akka-streams to reactive streams first, then from that to fs2.

**Comment by notxcain** [comment]
@Daenyth I think it's the same amount of indirection, we just replace Akka Queue level with Reactive Streams level. Benchmarks would be great to have, indeed.

**Comment by notxcain** [comment]
Well, it's not as straightforward as I thought it would be :)

**Comment by krasserm** [comment]
No benchmarks, unfortunately ...

> Well, it's not as straightforward as I thought it would be :)

@notxcain can you elaborate? Would it still reduce complexity in the streamz converter module or just create new complexity somewhere else. I didn't look into the FS2 - RS integration yet but naively thought that building on top of it would off-load most of the integration complexity to FS2 as Akka Streams already has an RS interface i.e. is an RS implementation. Also, RS advertises that implementations are easy to connect (I never tried though), so could this even make converter module obsolete?

**Comment by notxcain** [comment]
@krasserm The hard part is converting `Pipe` to `Flow` and vice versa. Everything else seems to be easy. If there is time, I'll give it another shot tomorrow.

**Comment by krasserm** [comment]
@notxcain I think conversion of sources and sinks is actually more important. Would be really interesting to see an implementation of it! 

Have you and/or @Daenyth ever used the Pipe <-> Flow conversion? I never did, just added it for completeness. Would you see it as a problem to drop support for that?

**Comment by Daenyth** [comment]
Yes, I've used it. I *probably* don't need it, but would prefer to keep it if possible.

It's avoidable by doing
```scala
val p: fs2.Pipe...
val as: Source[A, NU]
val snk: AkkaSink[A, NU]

as.toStream[IO].through(p).toSource().runWith(snk)
```
and vice-versa. But it's not as nice.

**Comment by krasserm** [comment]
@Daenyth sure, then let's keep it. Maybe it makes sense to let the existing converter implementation co-exist with another one based on FS2-RS and then see if we can deprecate the current implementation at some time.

**Comment by notxcain** [comment]
@krasserm I think that’s it.

**Comment by krasserm** [comment]
Thanks a lot @notxcain, great contribution!

**Comment by notxcain** [comment]
@krasserm my pleasure 😃

**Pull request merged**